### PR TITLE
Handle bounds properly when grid tiles crosses the dateline

### LINKED ISSUE #83299: [CI] GeoHexAggregatorTests testBounds failing
**Build scan:**
https://gradle-enterprise.elastic.co/s/hpf3zg34qqtbs/tests/:x-pack:plugin:spatial:test/org.elasticsearch.xpack.spatial.search.aggregations.bucket.geogrid.GeoHexAggregatorTests/testBounds

**Reproduction line:**
`./gradlew ':x-pack:plugin:spatial:test' --tests "org.elasticsearch.xpack.spatial.search.aggregations.bucket.geogrid.GeoHexAggregatorTests.testBounds" -Dtests.seed=13373AEC80D5202F -Dtests.locale=pt -Dtests.timezone=America/Antigua -Druntime.java=17`

**Applicable branches:**
master

**Reproduces locally?:**
Yes

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.xpack.spatial.search.aggregations.bucket.geogrid.GeoHexAggregatorTests&tests.test=testBounds

**Failure excerpt:**
```
java.lang.AssertionError: (No message provided)

  at __randomizedtesting.SeedInfo.seed([13373AEC80D5202F:E6E09F5E8C9139C]:0)
  at org.junit.Assert.fail(Assert.java:86)
  at org.junit.Assert.assertTrue(Assert.java:41)
  at org.junit.Assert.assertFalse(Assert.java:64)
  at org.junit.Assert.assertFalse(Assert.java:74)
  at org.elasticsearch.search.aggregations.bucket.geogrid.GeoGridAggregatorTestCase.lambda$testBounds$14(GeoGridAggregatorTestCase.java:283)
  at org.elasticsearch.search.aggregations.bucket.geogrid.GeoGridAggregatorTestCase.testCase(GeoGridAggregatorTestCase.java:333)
  at org.elasticsearch.search.aggregations.bucket.geogrid.GeoGridAggregatorTestCase.testCase(GeoGridAggregatorTestCase.java:300)
  at org.elasticsearch.search.aggregations.bucket.geogrid.GeoGridAggregatorTestCase.testBounds(GeoGridAggregatorTestCase.java:274)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(NativeMethodAccessorImpl.java:-2)
  at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
  at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
  at java.lang.reflect.Method.invoke(Method.java:568)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:44)
  at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:375)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:824)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:475)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:375)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:831)
  at java.lang.Thread.run(Thread.java:833)

```

### PR DESCRIPTION
**Pull request by iverase**
There are situation where a grid tile produced by a GeoGridAggregation might cross the dateline. With hex tile is expected but it might happen with geohash as well. In this case, we are not handling properly the tile intersection with the bounds provided by the query.

This PR adds logic to handle those cases in Geohash and Geohex Grid aggregations.

fixes https://github.com/elastic/elasticsearch/issues/83299

### CODE REVIEW

**Comment by iverase** [comment]
@elasticmachine update branch

**Comment by iverase** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoHashCellIdSource.java [hunk 1]
```diff
@@ -85,21 +85,42 @@ protected int advanceValue(org.elasticsearch.common.geo.GeoPoint target, int val
...
                 values[valuesIdx] = Geohash.longEncode(hash);
                 return valuesIdx + 1;
             }
             return valuesIdx;
         }
 
         private boolean validHash(String hash) {
-            final Rectangle rectangle = Geohash.toBoundingBox(hash);
+            final Rectangle rect = Geohash.toBoundingBox(hash);
+            if (rect.getMaxX() < rect.getMinX()) {
```
Comment (imotov) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoHashCellIdSource.java:97]: This whole thing could really use a drawing in the comments 😄 .

**Review by imotov — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoHashCellIdSource.java (hunk 1 above)
Comment (iverase) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoHashCellIdSource.java:100]: done

**Pull request merged**